### PR TITLE
Tool get_gpu_info (using lshw, lspci) with test

### PR DESCRIPTION
**Pull request by jack-lascala**
Implementation of a small tool to automate GPU architecture detection.

### CODE REVIEW

File: devito/archinfo.py [hunk 1]
```diff
@@ -117,6 +118,121 @@ def get_cpu_brand():
...
+    #   otherwise None
+    def homogenise_gpus(gpus):
+        if gpu_infos == []:
+            # warning('No graphics cards detected')
+            return None
+
+        if all(dev == gpu_infos[0] for dev in gpu_infos):
+            gpu_infos[0]['ncards'] = len(gpu_infos)
+            return gpu_infos[0]
+        # warning('Different models of graphics cards detected, should be homogenous')
```
Comment (FabioLuporini) [inline devito/archinfo.py:201]: commented out warning?

File: devito/archinfo.py [hunk 2]
```diff
@@ -117,6 +118,121 @@ def get_cpu_brand():
...
+
+        # Parse the information for all the devices listed with lspci
+        gpu_infos = lspci_gpu_info(raw_info)
+        return homogenise_gpus(gpu_infos)
+
+    except OSError:
+        pass
+
+    # If all calls to supported tools fail
+    # warning('No supported tool for GPU detection available')
```
Comment (FabioLuporini) [inline devito/archinfo.py:232]: commented out warning?

File: devito/archinfo.py (hunk 1 above)
Comment (jack-lascala) [reply devito/archinfo.py]: Re-introduced warning in newer commit

File: devito/archinfo.py [hunk 3]
```diff
@@ -1,13 +1,15 @@
 """Collection of utilities to detect properties of the underlying architecture."""
 
-from subprocess import PIPE, Popen
+from subprocess import PIPE, Popen, DEVNULL
 
 import numpy as np
 import cpuinfo
 import psutil
 
+from cached_property import cached_property
```
Comment (FabioLuporini) [inline devito/archinfo.py:9]: this one needs to be together with the third-party imports, such as numpy, cpuinfo, psutil, not with the devito imports

File: devito/archinfo.py [hunk 4]
```diff
@@ -1,13 +1,15 @@
...
 from devito.logger import warning
 from devito.tools.memoization import memoized_func
+from devito.tools.utils import all_equal
```
Comment (FabioLuporini) [inline devito/archinfo.py:12]: merge this line with the previous one, and ensure alphabetical ordering, ie

```
from devito.tools import all_equal, memoized_func
```

File: devito/archinfo.py [hunk 5]
```diff
@@ -117,6 +119,121 @@ def get_cpu_brand():
     return cpu_info
 
 
+@memoized_func
+def get_gpu_info():
+    """Attempt GPU info autodetection."""
+
+    # Filters out virtual GPUs from a list of GPU dictionaries
```
Comment (FabioLuporini) [inline devito/archinfo.py:126]: First person typically, so "Filter out ..."

File: devito/archinfo.py [hunk 6]
```diff
@@ -117,6 +119,121 @@ def get_cpu_brand():
...
+        # Separate the output into lines for processing
+        lines = text.replace('\\n', '\n')
+        lines = lines.splitlines()
+
+        gpu_infos = []
+        for line in lines:
+            # Graphics cards are listed as VGA or 3D controllers in lspci
+            if 'VGA' in line or '3D' in line:
+                gpu_info = {}
+                gpu_info['product'] = line.split(':')[1].split(':')[1].split(':').lstrip()
```
Comment (FabioLuporini) [inline devito/archinfo.py:184]: that doesn't look very robust... can you add a comment at least?

Comment (FabioLuporini) [reply devito/archinfo.py:184]: and an example of one such parsed line?

File: devito/archinfo.py [hunk 7]
```diff
@@ -117,6 +119,121 @@ def get_cpu_brand():
...
+
+                if '[' in line:
+                    gpu_info['architecture'] = line.split('[')[1].split(']')[0]
```
Comment (FabioLuporini) [inline devito/archinfo.py:187]: I wonder whether here, and actually in a few other places in this funcion, you should actually use a regex

File: devito/archinfo.py (hunk 6 above)
Comment (FabioLuporini) [reply devito/archinfo.py:184]: you probably need to use a regex. See comment below

File: devito/archinfo.py [hunk 8]
```diff
@@ -117,6 +119,121 @@ def get_cpu_brand():
...
+    #   otherwise None
+    def homogenise_gpus(gpus):
+        if gpu_infos == []:
+            warning('No graphics cards detected')
+            return None
+
+        if all_equal(gpu_infos):
+            gpu_infos[0]['ncards'] = len(gpu_infos)
+            return gpu_infos[0]
+        warning('Different models of graphics cards detected')
```
Comment (FabioLuporini) [inline devito/archinfo.py:204]: nitpicking: I'd leave a blank line before and after this warning

**Pull request merged**